daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala

```diff
@@ -151,9 +153,77 @@ object SValue {
       s"SPAP($prim, ${actuals.asScala.mkString("[", ",", "]")}, $arity)"
   }
 
+  /** We split SRecord (interface) from SRecordRep (implementation/representation)
+    *
+    * The interface supports creation from separate arrays of fields and values
+    * This is used throughout test code.
+    *
+    * The interface also supports (via unapply) the (legacy) reverse deconstruction.
+    * This is used by daml-script.
+    *
+    * The implementation/representation is via a scala Map.
+    * This representation is simple to manipulate (lookupField/updateField).
+    *
+    * This representation makes illegal cases unrepresentable -- i.e.
+    * - mismatched counts of fields/values
+    * - repeated field names
+    * And prevent brittle access to element values via indexing.
+    *
+    * Also note updateField has logarithmic complexity (where N is the number of fields)
+    * rather than the linear complexity that an array of element values would entail.
+    *
+    * The representation also includes an ordered field list.
+    * This is needed to support the legacy interface used by daml-script.
+    * And is also used when we convert the svalue back to a normalised LF value.
+    */
+
+  sealed abstract class SRecord extends SValue {
+    def id: Identifier
+    def fields: ImmArray[Name]
+    def values: util.ArrayList[SValue]
+    def lookupField(field: Name): SValue
+    def updateField(field: Name, value: SValue): SRecord
+  }
+
+  object SRecord {
+    def apply(id: Identifier, fields: ImmArray[Name], values: util.ArrayList[SValue]): SRecord = {
+      assert(fields.length == values.size)
+      val zipped = (fields.toSeq zip values.asScala).toList.reverse
+      SRecordRep(id, fields, zipped.toMap)
+    }
+    def unapply(x: SRecord): Option[(Identifier, ImmArray[Name], util.ArrayList[SValue])] = {
+      Some((x.id, x.fields, x.values))
+    }
+  }
+
   @SuppressWarnings(Array("org.wartremover.warts.ArrayEquals"))
-  final case class SRecord(id: Identifier, fields: ImmArray[Name], values: util.ArrayList[SValue])
-      extends SValue
+  final case class SRecordRep(id: Identifier, fields: ImmArray[Name], dict: Map[Name, SValue])
+      extends SRecord {
+
+    def lookupField(field: Name): SValue = {
+      dict.get(field) match {
+        case Some(v) => v
+        case None =>
+          throw SError.SErrorCrash(
+            NameOf.qualifiedNameOfCurrentFunc,
+            s"SRecord.lookupField($field) : no such field, have:[$fields]",
+          )
+      }
+    }
+
+    def updateField(field: Name, value: SValue): SRecord = {
+      this.copy(dict = dict + (field -> value))
+    }
+
+    def values: util.ArrayList[SValue] = {
+      val res = new util.ArrayList[SValue]()
+      fields.foreach { case field =>
+        val value = this.lookupField(field)
+        discard[Boolean](res.add(value))
+      }
```

Comment:
Maybe just `fields.map(this.lookupField(_))`?

Reply:
No, sadly this is horrible imperative code (`.add`)) to construct a `util.ArrayList`
I don't know if there is a way to do this more functionally,

Reply:
avoid `toList`. 
```suggestion 
fields.toSeq.view.map(this.lookupField).to(ArrayList)
```

Reply:
See #17106
Turns out the `.view` was not needed. Just the `.toSeq` was enough.